### PR TITLE
Feature/command palette

### PR DESCRIPTION
**Pull request by Akin909**
Add initial implementation of command palette
palette

currently:
* hooks up input handling to either be handled by Oni or passed to Neovim
* Reads commands out of a bindings list (later to be derived from a `keybindings.json` file

### Outstanding

- [x] Hook up handler for commands on select
- [x] Setup state for managing when input control is in oni or neovim, e.g. whilst the command palette is open it should *own* user input
- [x] hook up `keybindings.json` (maybe separate PR)

### CODE REVIEW

File: src/editor/Core/CommandPalette.re [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+let reduce = (state: t, action: Actions.t) =>
+  switch (action) {
+  | CommandPalettePosition(pos) => {
+      ...state,
+      selectedItem: position(state.selectedItem, pos, state.commands),
+    }
+  | CommandPaletteOpen => {...state, isOpen: true}
+  | CommandPaletteClose => {...state, isOpen: false}
+  | _ => state
```
Comment (Akin909) [inline src/editor/Core/CommandPalette.re]: this pattern removes `exhaustiveness checking` which is very sad :sob:

File: assets/configuration/keybindings.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (bryphe) [inline assets/configuration/keybindings.json]: Woah, nice! We have keybindings now! 😍

File: src/editor/Core/CommandPalette.re [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  | CommandPaletteStart(commands) => {...state, commands}
+  | CommandPalettePosition(pos) => {
+      ...state,
+      selectedItem: position(state.selectedItem, pos, state.commands),
+    }
+  | CommandPaletteOpen => {...state, isOpen: true}
+  | CommandPaletteClose => {...state, isOpen: false}
+  | CommandPaletteSelect =>
+    /**
+       TODO: Refactor this to middleware so this action is handled like a redux side-effect
```
Comment (bryphe) [inline src/editor/Core/CommandPalette.re]: Thanks for thinking about this 👍

File: src/editor/Core/CommandPalette.re [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  switch (effects) {
+  | Some(e) => {
+      isOpen: false,
+      commands: commandPaletteCommands(e),
+      selectedItem: 0,
+    }
+  | None => {isOpen: false, commands: [], selectedItem: 0}
+  };
+
+let make = (~effects: Effects.t) =>
```
Comment (bryphe) [inline src/editor/Core/CommandPalette.re]: I like your thinking here on how to make the command actions 'pure' 👍

File: src/editor/Core/Commands.re [hunk 5]
```diff
@@ -0,0 +1,43 @@
+type oniCommand = {
+  name: string,
+  command: unit => list(Actions.t),
+};
+
+type t = list(oniCommand);
+
+let oniCommands = [
+  {
+    name: "commandPalette.open",
```
Comment (bryphe) [inline src/editor/Core/Commands.re]: Thanks for starting up the `Commands` concept. It's a neat idea to have them map directly to our editor actions 👍

**Pull request merged**